expyfun/stimuli/_tracker.py

```diff
@@ -0,0 +1,858 @@
+        lists. The length of ``change_criteria`` must be the same as the
+        length of ``step_size_up`` and ``step_size_down``. See note below for
+        more specific usage information. Should be None if static step sizes
+        are used.
+    change_rule : str
+        Whether to change paramters based on 'trials' or 'reversals'.
+    x_min : float
+        The minimum value that the tracker level (``x``) is allowed to take.
+    x_max : float
+        The maximum value that the tracker level (``x``) is allowed to take.
+
+    Returns
+    -------
+    tracker : instance of TrackerUD
+        The up-down tracker object.
+
+    Notes
+    -----
+    It is common to use dynamic parameters in an adaptive tracker. For example:
+    the step size is often large for the first couple reversals to get in the
+    right general area, and then it is reduced for the remainder of the track.
+    This class allows that functionality by defining that parameter with a list
+    of values rather than a scalar. The parameter will change to the next value
+    in the list whenever a change criterion (number of trials or reversals) is
+    encountered. This means that the length of the list defining a dynamic
+    parameter must always be the same as that of ``change_criteria``, and the
+    first element of change_criteria must always be 0.
+    For the example given above:
+        ``step_size_up=[1., 0.2]``, ``step_size_down=[1., 0.2]``,
+        ``change_criteria=[0, 2]``, ``change_rule='reversals'``
+    would change the step sizes from 1 to 0.2 after two reversals.
+
+    If dynamic step sizes are used, both ``step_size_up`` and
+    ``step_size_down`` must have the same length as each other and
+    ``change_cirteria``. If static step sizes are used, both ``step_size_up``
+    and ``step_size_up`` must be scalars and ``change_criteria`` must be None.
+    """
+    def __init__(self, callback, up, down, step_size_up, step_size_down,
+                 stop_criterion, stop_rule, start_value, change_criteria=None,
+                 change_rule='reversals', x_min=None, x_max=None):
+        self._callback = check_callback(callback)
+        self._up = up
+        self._down = down
+        self._stop_criterion = stop_criterion
+        self._stop_rule = stop_rule
+        self._start_value = start_value
+        self._x_min = x_min
+        self._x_max = x_max
+
+        if change_criteria is None:
+            change_criteria = [0]
+        elif change_criteria[0] != 0:
+            raise(ValueError('First element of change points must be 0.'))
+        self._change_criteria = np.asarray(change_criteria)
+        if change_rule not in ['trials', 'reversals']:
+            raise(ValueError, "change_rule must be either 'trials' or "
+                  "'reversals'")
+        else:
+            self._change_rule = change_rule
+
+        if np.isscalar(step_size_up):
+            step_size_up = [step_size_up] * len(change_criteria)
```

Comment:
Not the same, because this will turn a scalar `step_size_up` into a list-like of the proper length. However, I don't like that functionality--it allows too much sloppiness. So I will change to what you suggest and require either ALL scalars or ALL list-like of the proper length.